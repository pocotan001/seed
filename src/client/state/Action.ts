// Flux standard action
// https://github.com/redux-utilities/flux-standard-action
export interface Action<Type, Payload = null, Meta = undefined> {
  type: Type;
  payload: Payload;
  meta?: Meta;
}

export interface ErrorAction<Type, Payload = Error, Meta = undefined> {
  type: Type;
  payload: Payload;
  error: true;
  meta?: Meta;
}
