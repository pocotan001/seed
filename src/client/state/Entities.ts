export interface NormalizedEntities<Entity> {
  [id: number]: Entity;
}
