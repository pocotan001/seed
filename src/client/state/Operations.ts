import { Dispatch } from "redux";
import { Persistor } from "redux-persist";
import { Repositories } from "../infra/repositories";
import createAuthOperations, { AuthOperations } from "./auth/AuthOperations";
import createLoadingOperations, {
  LoadingOperations
} from "./loading/LoadingOperations";
import createRootOperations, { RootOperations } from "./root/RootOperations";

export interface Operations {
  root: RootOperations;
  auth: AuthOperations;
  loading: LoadingOperations;
}

const createOperations = (
  dispatch: Dispatch,
  persistor: Persistor,
  repos: Repositories
): Operations => ({
  root: createRootOperations(dispatch),
  auth: createAuthOperations(dispatch, persistor, repos.auth),
  loading: createLoadingOperations(dispatch)
});

export default createOperations;
