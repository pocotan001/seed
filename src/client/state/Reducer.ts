import { connectRouter } from "connected-react-router";
import { History } from "history";
import { combineReducers, Reducer as ReduxReducer } from "redux";
import { reducer as formReducer } from "redux-form";
import { persistReducer } from "redux-persist";
import storage from "redux-persist/lib/storage";
import storageSession from "redux-persist/lib/storage/session";
import authReducer from "./auth/AuthReducer";
import loadingReducer from "./loading/LoadingReducer";
import { State } from "./State";

export type Reducer = ReduxReducer<State>;

const createReducer = (history: History): Reducer =>
  combineReducers({
    router: connectRouter(history),
    form: persistReducer(
      {
        key: "form",
        storage: storageSession,
        whitelist: []
      },
      formReducer
    ),
    auth: persistReducer(
      {
        storage,
        key: "auth"
      },
      authReducer
    ),
    loading: loadingReducer
  });

export default createReducer;
