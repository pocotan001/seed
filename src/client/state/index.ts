export { RootState as State } from "./root/RootState";
export { default } from "./root/RootReducer";
