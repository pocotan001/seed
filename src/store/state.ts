import { createLocation, Location } from "history";
import { extendObservable, observable } from "mobx";
import config from "~/config";
import { INormalizedEntities } from "~/domain/Normalized";

export enum SessionKey {
  EXAMPLE_FORM = "EXAMPLE_FORM"
}

export const defaultState: State = Object.freeze({
  app: {
    locationOrigin: "",
    hasError: false
  },
  history: {
    location: createLocation(""),
    visited: {}
  },
  head: {
    title: config.siteName,
    meta: [],
    link: []
  },
  loading: {
    percent: 0,
    hidden: true
  },
  session: {},
  entities: {
    user: null,
    cats: {}
  },
  results: {
    cats: {}
  },
  cats: {
    totalCount: 0
  }
});

export class State {
  app!: {
    locationOrigin: string;
    hasError: boolean;
  };

  history!: {
    location: Location;
    visited: { [key: string]: boolean };
  };

  head!: {
    title: string;
    meta: Array<React.MetaHTMLAttributes<HTMLMetaElement>>;
    link: Array<React.LinkHTMLAttributes<HTMLLinkElement>>;
  };

  loading!: {
    percent: number;
    hidden: boolean;
  };

  session!: { [K in SessionKey]?: any };

  entities!: INormalizedEntities;

  results!: Record<
    "cats",
    {
      [serializedParams: string]: string[];
    }
  >;

  cats!: {
    totalCount: number;
  };

  constructor(initialState?: Partial<State>) {
    extendObservable(
      this,
      { ...defaultState, ...initialState },
      {},
      {
        defaultDecorator: observable.shallow
      }
    );
  }
}

const createState = (initialState?: Partial<State>) => new State(initialState);

export default createState;
