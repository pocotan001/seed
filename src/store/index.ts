import AppStore from "./AppStore";
import AuthenticationStore from "./AuthenticationStore";
import CatStore from "./CatStore";
import HeadStore from "./HeadStore";
import HistoryStore from "./HistoryStore";
import LoadingStore from "./LoadingStore";
import SessionStore from "./SessionStore";
import { State } from "./state";
import { IStoreContext } from "./Store";

export class RootStore {
  state: State;
  app: AppStore;
  auth: AuthenticationStore;
  head: HeadStore;
  history: HistoryStore;
  loading: LoadingStore;
  session: SessionStore;
  cat: CatStore;

  constructor(state: State, ctx: IStoreContext) {
    this.state = state;
    this.app = new AppStore(state, ctx);
    this.auth = new AuthenticationStore(state, ctx);
    this.head = new HeadStore(state, ctx);
    this.history = new HistoryStore(state, ctx);
    this.loading = new LoadingStore(state, ctx);
    this.session = new SessionStore(state, ctx);
    this.cat = new CatStore(state, ctx);
  }
}

const createStore = (state: State, ctx: IStoreContext) =>
  new RootStore(state, ctx);

export default createStore;
