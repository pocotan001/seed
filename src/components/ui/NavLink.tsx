import { inject, observer } from "mobx-react";
import * as React from "react";
import { matchPath } from "react-router";
import { RootStore } from "~/store";
import Link, { ILinkProps } from "./Link";

interface INavLinkProps extends ILinkProps {
  activeClassName: string;
  path?: string;
  exact?: boolean;
  strict?: boolean;
  store?: RootStore;
}

const NavLink: React.SFC<INavLinkProps> = ({
  href,
  activeClassName,
  className,
  path,
  exact,
  strict,
  store,
  ...rest
}) => {
  const isActive = Boolean(
    matchPath(store!.state.history.location.pathname, {
      path: path || href,
      exact,
      strict
    })
  );

  const computedClassName = isActive
    ? [className, activeClassName].filter(Boolean).join(" ")
    : className;

  return (
    <Link href={href} className={computedClassName} store={store} {...rest} />
  );
};

export default inject("store")(observer(NavLink));
