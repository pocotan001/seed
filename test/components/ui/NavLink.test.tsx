import * as React from "react";
import * as renderer from "react-test-renderer";
import NavLink from "~/components/ui/NavLink";
import { RootStore } from "~/store";
import { defaultState } from "~/store/state";

let store: RootStore;

describe("<NavLink>", () => {
  beforeEach(() => {
    store = {
      state: { ...defaultState },
      history: {
        push: () => undefined
      }
    } as any;
  });

  it("renders correctly", () => {
    const tree = renderer
      .create(
        <NavLink href="/a" activeClassName="a" store={store}>
          alo
        </NavLink>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("changes the class name when actived", () => {
    store.state.history.location.pathname = "/a";

    const tree = renderer
      .create(
        <NavLink href="/a" activeClassName="a" store={store}>
          alo
        </NavLink>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
