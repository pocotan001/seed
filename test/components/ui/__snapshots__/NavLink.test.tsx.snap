// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavLink> changes the class name when actived 1`] = `
<a
  className="a sc-bdVaJa kSKvLH"
  href="/a"
  onClick={[Function]}
>
  alo
</a>
`;

exports[`<NavLink> renders correctly 1`] = `
<a
  className="sc-bdVaJa kSKvLH"
  href="/a"
  onClick={[Function]}
>
  alo
</a>
`;
